src/spatialdata/_core/spatialdata.py

```diff
@@ -537,37 +655,93 @@ def filter_by_coordinate_system(self, coordinate_system: str | list[str], filter
+            filter_tables is also set to True.
 
         Returns
         -------
         The filtered SpatialData.
         """
-        from spatialdata._core.query.relational_query import _filter_table_by_coordinate_system
+        # TODO: decide whether to add parameter to filter only specific table.
+
         from spatialdata.transformations.operations import get_transformation
 
         elements: dict[str, dict[str, SpatialElement]] = {}
         element_paths_in_coordinate_system = []
         if isinstance(coordinate_system, str):
             coordinate_system = [coordinate_system]
         for element_type, element_name, element in self._gen_elements():
-            transformations = get_transformation(element, get_all=True)
-            assert isinstance(transformations, dict)
-            for cs in coordinate_system:
-                if cs in transformations:
-                    if element_type not in elements:
-                        elements[element_type] = {}
-                    elements[element_type][element_name] = element
-                    element_paths_in_coordinate_system.append(element_name)
-
-        if filter_table:
-            table = _filter_table_by_coordinate_system(self.table, element_paths_in_coordinate_system)
+            if element_type != "tables":
+                transformations = get_transformation(element, get_all=True)
+                assert isinstance(transformations, dict)
+                for cs in coordinate_system:
+                    if cs in transformations:
+                        if element_type not in elements:
+                            elements[element_type] = {}
+                        elements[element_type][element_name] = element
+                        element_paths_in_coordinate_system.append(element_name)
+        tables = self._filter_tables(
+            filter_tables, "cs", include_orphan_tables, element_paths=element_paths_in_coordinate_system
+        )
+
+        return SpatialData(**elements, tables=tables)
+
+    def _filter_tables(
+        self,
+        filter_tables: bool = True,
+        by: Literal["cs", "elements"] | None = None,
+        include_orphan_tables: bool = False,
+        element_paths: str | list[str] | None = None,
+        elements_dict: dict[str, dict[str, Any]] | None = None,
+    ) -> Tables | dict[str, AnnData]:
+        """
+        Filter tables by coordinate system or elements or return tables.
+
+        Parameters
+        ----------
+        filter_tables
+            If True (default), the tables will be filtered to only contain regions
+            of an element belonging to the specified coordinate system(s).
+        by
+            Filter mode. Valid values are "cs" or "elements". Default is None.
+        include_orphan_tables
+            Flag indicating whether to include orphan tables. Default is False.
+        element_paths
+            Check whether this should be changed into element names in coordinate system
```

Comment:
yes, it's to be changed as discussed thanks

Reply:
resolved with refactor to `filter_by_element_names`.